BUG/REF: TimedeltaIndex.__new__

jbrockmendel opened the pull request:
`TimedeltaIndex.__new__` type-checking is tough to decipher, and hides multiple bugs

```
>>> dti = pd.date_range('2016-01-01', periods=3)
>>> pd.TimedeltaIndex(np.array(dti))
TimedeltaIndex(['16801 days', '16802 days', '16803 days'], dtype='timedelta64[ns]', freq=None)
```

In addition to this incorrect behavior being supported, it is actually _tested_ in `tests.indexes.timedeltas.test_ops` in what I have to assume is a copy/paste mixup of some kind. 

```
>>> pd.TimedeltaIndex([0.5])
TimedeltaIndex(['0 days'], dtype='timedelta64[ns]', freq=None)
```

This PR fixes the constructor to

1) be much more linear and clear than the existing constructor
2) reject datetime64 and anything that isn't explicitly supported
3) accept floating-dtype if and only if they can be converted losslessly: `np.array([19.0])` is OK, `np.array([19.5])` is not.  `np.nan` is correctly converted to `NaT`

Other assorted small changes:
Move an incorrectly placed `# ----` section divider in scalar timedelta tests.
Change a repeated-4-times 4-liner (setting `result.freq = inferred_freq`) to a 2-linear
Move raising and returning-early cases to the top of `DatetimeIndex.__new__` and `TimedeltaIndex.__new__`

Has tests, does not have whatsnew note, pending suggestions as to which parts merit notes.

Does not significantly change the TimedeltaArray constructor; in an upcoming step we can move much of the logic there and have the TDI constructor wrap a/the TDA constructor.

jbrockmendel committed:
simplify+fix+test TimedeltaIndex constructor

jbrockmendel committed:
tests for datetime64 data being invalid, floats being valid iff non-lossy

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into pre-fixes

jbrockmendel committed:
comments and whitespace

pep8speaks commented:
Hello @jbrockmendel! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/arrays/datetimes.py`](https://github.com/jbrockmendel/pandas/blob/1ff432b1f6927812c5fe0bfbb98b2bf594c8fd14/pandas/core/arrays/datetimes.py) !

 - There are no PEP8 issues in the file [`pandas/core/arrays/timedeltas.py`](https://github.com/jbrockmendel/pandas/blob/1ff432b1f6927812c5fe0bfbb98b2bf594c8fd14/pandas/core/arrays/timedeltas.py) !

 - There are no PEP8 issues in the file [`pandas/core/indexes/datetimes.py`](https://github.com/jbrockmendel/pandas/blob/1ff432b1f6927812c5fe0bfbb98b2bf594c8fd14/pandas/core/indexes/datetimes.py) !

 - There are no PEP8 issues in the file [`pandas/core/indexes/timedeltas.py`](https://github.com/jbrockmendel/pandas/blob/1ff432b1f6927812c5fe0bfbb98b2bf594c8fd14/pandas/core/indexes/timedeltas.py) !

 - There are no PEP8 issues in the file [`pandas/tests/arithmetic/test_timedelta64.py`](https://github.com/jbrockmendel/pandas/blob/1ff432b1f6927812c5fe0bfbb98b2bf594c8fd14/pandas/tests/arithmetic/test_timedelta64.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/timedeltas/test_construction.py`](https://github.com/jbrockmendel/pandas/blob/1ff432b1f6927812c5fe0bfbb98b2bf594c8fd14/pandas/tests/indexes/timedeltas/test_construction.py) !

 - There are no PEP8 issues in the file [`pandas/tests/indexes/timedeltas/test_ops.py`](https://github.com/jbrockmendel/pandas/blob/1ff432b1f6927812c5fe0bfbb98b2bf594c8fd14/pandas/tests/indexes/timedeltas/test_ops.py) !

 - There are no PEP8 issues in the file [`pandas/tests/scalar/timedelta/test_arithmetic.py`](https://github.com/jbrockmendel/pandas/blob/1ff432b1f6927812c5fe0bfbb98b2bf594c8fd14/pandas/tests/scalar/timedelta/test_arithmetic.py) !

jbrockmendel committed:
GH references

gfyoung commented on pandas/core/arrays/datetimes.py:
```diff
@@ -221,9 +221,7 @@ def __new__(cls, values, freq=None, tz=None, dtype=None):
-            inferred = result.inferred_freq
-            if inferred:
-                result.freq = to_offset(inferred)
+            result.freq = to_offset(result.inferred_freq)
```
Just curious, why can we drop the `if`-logic?

jbrockmendel replied:
because it is unnecessary.  if `inferred` is `None`, then we're setting `result.freq=None` which is what it should be anyway.  It isn't a big deal, but these 4 lines are repeated 4 times and after looking at these for weeks I find it an eyesore.

gfyoung commented on pandas/core/indexes/timedeltas.py:
```diff
@@ -154,32 +153,73 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
+            else:
+                data = data.view(_TD_DTYPE)
+        elif is_float_dtype(data):
+            # We allow it if and only if it can be converted lossessly
```
lossessly --> losslessly

gfyoung commented on pandas/tests/indexes/timedeltas/test_construction.py:
```diff
@@ -10,6 +10,30 @@
+            TimedeltaIndex(dti)
+
+        with pytest.raises(TypeError):
+            TimedeltaIndex(np.asarray(dti))
```
Let's check error message, if possible.

jbrockmendel replied:
will do

gfyoung commented on pandas/core/indexes/timedeltas.py:
```diff
@@ -154,32 +153,73 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
+        if not hasattr(data, 'dtype'):
+            # e.g. list, tuple
+            if np.ndim(data) == 0:
+                # i.e.g generator
```
i.e.g ?

jbrockmendel replied:
typo, thanks

gfyoung commented on pandas/core/indexes/timedeltas.py:
```diff
@@ -154,32 +153,73 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
+        # Convert whatever we have into timedelta64[ns] dtype
+        if is_object_dtype(data) or is_string_dtype(data):
+            # no need to make a copy, need to convert if string-dtyped
+            data = np.array(data, dtype=np.object_, copy=False)
```
Can we use `np.asarray` here?

jbrockmendel replied:
it should be specifically an np.ndarray

gfyoung commented on pandas/core/indexes/timedeltas.py:
```diff
@@ -154,32 +153,73 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
+                            .format(dtype=data.dtype, cls=cls.__name__))
 
-        assert data.dtype == 'm8[ns]', data.dtype
+        data = np.array(data, copy=copy)
```
Can we use np.asarray here?

jbrockmendel replied:
same as above, this needs to be specifically an np.ndarray

jorisvandenbossche commented:
@jbrockmendel It seems you are adding net a lot of code. Is that because some of the things were previously handled by the `to_timedelta` call?

But that said, I think some of those things (eg the float) handling should actually be fixed in `to_timedelta`, and not here, and the `TimedeltaIndex` constructor can keep making use of that.

jorisvandenbossche commented:
> Does not significantly change the TimedeltaArray constructor; in an upcoming step we can move much of the logic there and have the TDI constructor wrap a/the TDA constructor.

@jbrockmendel as also mentioned in the other PR, Tom and I assumed that the consensus was to *not* do this (keep the TDA constructor simple), so if you still want to do this, please respond to the issue I linked then: https://github.com/pandas-dev/pandas/issues/23212

jorisvandenbossche commented:
I would personally be fine with raising a TypeError, since I think this can be quite surprising behaviour (it's just using the integer values). But also fine to not deviate from numpy behaviour. 

But given it is numpy behaviour, so if we want to change it, I think we should deprecate it first.

TomAugspurger commented:
> IMO this is a TypeError

"is a" -> "should be"?

I don't really disagree, and we *do* raise for `TimedeltaIndex(DatetimeIndex)`, so some consistency there would be nice. But I suspect that we have people relying on this, so a deprecation cycle would be needed.

jreback commented:
ok by me to deprecate first

TomAugspurger commented:
> this is really preventing foot shooting

I do realize my argument of "but NumPy allows it" isn't great :)

```python
In [3]: np.array([1, np.nan]).astype(int)
Out[3]: array([                   1, -9223372036854775808])
```

jbrockmendel commented on pandas/core/indexes/datetimes.py:
```diff
@@ -239,6 +239,21 @@ def __new__(cls, data=None,
+
+        if is_scalar(data):
+            raise ValueError("{cls}() must be called with a "
+                             "collection of some kind, {data} was passed"
```
We'll change that before long.  Right now I'm just trying to pull one tooth at a time.

TomAugspurger replied:
Does changing this require changing existing tests?

TomAugspurger replied:
Ah, I see that it was just moved, so probably.

jreback replied:
pls fix this

jbrockmendel replied:
Not in this PR.

jbrockmendel replied:
never mind, will do

jbrockmendel commented on pandas/core/indexes/timedeltas.py:
```diff
@@ -154,32 +153,73 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
+
+        # Convert whatever we have into timedelta64[ns] dtype
+        if is_object_dtype(data) or is_string_dtype(data):
+            # no need to make a copy, need to convert if string-dtyped
```
because we have several tests that specifically pass e.g. `np.array(['1 days', '4 days'], dtype='<S6')`.  (BTW we don't have similar tests for DatetimeIndex, which I intend to fix in the DTI analogue of this PR)

jbrockmendel commented on pandas/core/indexes/timedeltas.py:
```diff
@@ -154,32 +153,73 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
+            mask = np.isnan(data)
+            casted = data.astype(np.int64)
+            if not (casted[~mask] == data[~mask]).all():
+                raise TypeError("floating-dtype data cannot be losslessly "
```
will take a look

jbrockmendel commented:
@jorisvandenbossche 

> But that said, I think some of those things (eg the float handling) should actually be fixed in to_timedelta, and not here,

It can be difficult to tell exactly what is handled within the to_timedelta calls because there are two of them reached under separate conditions, and most of the relevant logic is then done again below that anyway.  I'd be +1 on improving to_timedelta later.

> as also mentioned in the other PR, Tom and I assumed that the consensus was to not do this (keep the TDA constructor simple), so if you still want to do this, please respond to the issue I linked then: #23212

I'll restate my opinion there when I'm ready, thank you.  I avoided changing the Array constructors specifically to avoid this being an issue; really shouldn't have mentioned it all in the OP.

jbrockmendel commented:
> I think the important question is whether we want [...] By "constructor" do you mean TimedeltaArray.__init__, or just a generic term for creating a TimedeltaArray from some input?

@TomAugspurger same as above; that is _an_ important question, but I went out of my way to keep it totally orthogonal to this PR.

jbrockmendel commented:
It sounds like the discussion w/r/t `TimedeltaIndex(np.array(dti))` has resolved itself overnight with an agreement that there should be a deprecation before a TypeError.  Am I reading the room correctly?

TomAugspurger commented:
> It sounds like the discussion w/r/t TimedeltaIndex(np.array(dti)) has resolved itself overnight with an agreement that there should be a deprecation before a TypeError. Am I reading the room correctly?

I think so.

> I'll restate my opinion there when I'm ready, thank you. I avoided changing the Array constructors specifically to avoid this being an issue; really

Ok. I think I saw "refactor `TimedeltaIndex.__new__`" and thought it meant something closer to what we did for `PeriodIndex.__new__`, but we may not be there yet.

jorisvandenbossche commented:
> It can be difficult to tell exactly what is handled within the to_timedelta calls because there are two of them reached under separate conditions, and most of the relevant logic is then done again below that anyway. I'd be +1 on improving to_timedelta later.

The bug you are mentioning (`pd.TimedeltaIndex([0.5])` ignoring the decimal) is coming from `to_timedelta` doing that (I just verified it is indeed taking this path). So you are repeating and fixing here what `to_timedelta` is already doing, so if you want to fix this behaviour, it really seems logical to me to fix this in `to_timedelta` instead of expanding the code here.

> It sounds like the discussion w/r/t TimedeltaIndex(np.array(dti)) has resolved itself overnight with an agreement that there should be a deprecation before a TypeError. Am I reading the room correctly?

Yes, I think so.

jbrockmendel commented:
> So you are repeating and fixing here what to_timedelta is already doing, so if you want to fix this behaviour, it really seems logical to me to fix this in to_timedelta instead of expanding the code here.

AFAICT the to_timedelta call (L168 in master) is _almost_ superfluous given what comes below (which was recently moved there IIRC because it was bizarrely located in `_simple_new` before that).  If we disable the branch that calls `to_timedelta`, the only other changes we need to make everything work (or specifically, pass existing tests) is to a) convert generators to lists before passing into `data = np.array(data, copy=False)` and b) catch string-dtypes in the same block that catches `np.object_` just below that.  Which is to say: calling `to_timedelta` at all is highly redundant.

That said, an argument that I _would_ find compelling is that we definitely don't want the behavior of `TimedeltaIndex([x, y, z])` to diverge from the behavior of `to_timedelta([x, y, z])`, which this PR currently does.  So I'll take a look at combining these after some more caffeine.

jorisvandenbossche commented:
>  If we disable the branch that calls to_timedelta, the only other changes we need to make everything work (or specifically, pass existing tests) is to a) convert generators to lists before passing into data = np.array(data, copy=False) and b) catch string-dtypes in the same block that catches np.object_ just below that. Which is to say: calling to_timedelta at all is highly redundant.

If you are going to look at it, I would recommend trying to make the checks you mention here redundant, by letting `to_timedelta` handle it. Eg strings is something that `to_timedelta` is made for to handle. 
I think the object dtype handling could be removed, see below (but certainly didn't check all cases, so this might be too optimistic)


Dived into the (current master) code for a moment, and you have this:

https://github.com/pandas-dev/pandas/blob/28a42da41ca8e13efaa2ceb3939e576d08c232c8/pandas/core/indexes/timedeltas.py#L165-L173

Here, the conversion of everything that is not yet a timedelta64 array is first handled with `to_timedelta`. But at the end of the snippet, we still call `array_to_timedelta64(data)` (what is used inside `to_timedelta`) if the data is object dtype. But didn't we already just convert everything not m8 with `to_timedelta`? 
So this happens if `data` is a TimedeltaArrayMixin (or at least one of the cases): TimedeltaArrayMixin still has a m8[ns] numpy dtype, so the `getattr(data, 'dtype', None) != _TD_DTYPE` is passed; so we then convert it with `np.array(data)` to an array of Timedelta objects, to then convert it back to m8 with `array_to_timedelta64`. 
So this is of course just a redundant back and forth conversion, that can be avoided with a check if `data` is a TimedeltaArrayMixin, and in that case get the underlying numpy array of it. 
(similar to what DatetimeIndex also does, it has a check `if isinstance(data, DatetimeArrayMixin)`)

jbrockmendel committed:
deprecate instead of raising for datetime64 dtypes

jbrockmendel commented:
@jorisvandenbossche I think we're in agreement on the relevant part which is that the status quo is redundant and unclear.  Give me some time to adjust the PR to see if we can reach something closer to optimal.

jbrockmendel committed:
implement sequence_to_td64ns, deprecate datetime64 data, add and test fastpath for inference/validation

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into pre-fixes

jbrockmendel commented:
Pretty major set of changes since the previous round:

- instead of doing all the inference/conversion inside `TimedeltaIndex.__new__`, it now occurs in `arrays.timedeltas.sequence_to_td64ns`
- `to_timedelta` now dispatches to `sequence_to_td64ns`
- more careful checks for at-most-one copy
- fastpath freq inference/validation (and tests for each new path that produces)
- passing datetime64 data is deprecated instead of raising

jbrockmendel committed:
catch warnings

jorisvandenbossche reviewed (commented):
Thanks for the update! Some quick general comments:

- I would move the conversion code to `core/tools/timedeltas.py`, where `to_timedelta` already lives (and that keeps `arrays.timedeltas` also a bit leaner)
- Can you add tests for the changed behaviours for `to_timedelta` as well (the floats, not sure if there is something else)
- Regarding the floats, I am wondering if we should also first deprecate this? As this might also be a hard break for people using it.

jorisvandenbossche commented:
> Regarding the floats, I am wondering if we should also first deprecate this? As this might also be a hard break for people using it.

Actually, I think we need to revert it, as it is actually a feature. It does work as expected with bigger units (we just ignore anything smaller as a nanosecond, which is the full decimal at nanosecond resolution, but not at higher resolution):

```
In [12]: pd.TimedeltaIndex([1.5, 2.5], unit='D')
Out[12]: TimedeltaIndex(['1 days 12:00:00', '2 days 12:00:00'], dtype='timedelta64[ns]', freq=None)
```

This is broken on this branch, which also means that it is actually not tested ..

jbrockmendel commented:
You make a good point on the floats.  I'll revert and add a test for the example you gave.

w/r/t tools.timedeltas, I think that is a distinctly poor place for this logic to live.  The logic for how we create TimedeltaArrays belongs with... TimedeltaArrays.  Especially w/r/t keeping a DAG-like dependency structure, this simplifies things.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into pre-fixes

jbrockmendel committed:
revert float changes, and tests

TomAugspurger commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -395,6 +403,163 @@ def f(x):
+            data = list(data)
+        data = np.array(data, copy=False)
+    elif isinstance(data, ABCSeries):
+        data = data._values
```
Could potentially just get the freq here for Series[timedelta64[ns]]. If `TimedeltaArray` were done, we could change the following `elif` to an `if` and re-use that. For now I think we could check for timedelta dtype and do `inferred_freq = data.dt.freq` (which I assume is cheaper than re-inferring the freq?)

jbrockmendel replied:
Yah, I'm looking forward to this.

> and do `inferred_freq = data.dt.freq`

One of the first things I tried to do around here was get DatetimeIndex/TimedeltaIndex to have a `.dt` dummy property returning `self` so we wouldn't have to check for Index vs. Series; gotta get around to making that happen.

TomAugspurger commented on pandas/tests/indexes/timedeltas/test_arithmetic.py:
```diff
@@ -453,10 +453,16 @@ def test_timedelta_ops_with_missing_values(self):
         s1 = pd.to_timedelta(Series(['00:00:01']))
         s2 = pd.to_timedelta(Series(['00:00:02']))
-        sn = pd.to_timedelta(Series([pd.NaT]))
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
```
Can we get this stacklevel correct? Or are we calling this from two places with different stacklevels?

jorisvandenbossche replied:
Possibly not for both `to_timedelta` and `DatetimeIndex`, but so let's choose and eg take the correct stacklevel for the case of an array to `to_timedelta`

jbrockmendel replied:
I lost patience pretty quickly while trying.  It wouldn't surprise me that much if it turned out to be really simple ex-post.

jorisvandenbossche replied:
Can you do this as a follow-up as well?

jbrockmendel commented:
> Does changing this require changing existing tests?

No, the error message is unchanged.

jbrockmendel commented:
Thoughts here?  I'd like to get a move on the DTA analogue of this PR.

jreback commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -395,6 +403,163 @@ def f(x):
+            # non-nano unit
+            # TODO: watch out for overflows
+            data = data.astype(_TD_DTYPE)
+            copy = False
```
what if this is not true?

jbrockmendel replied:
How do you mean?  Won't the `astype` always have to make a copy in this context?

jreback replied:
i don't know if it does or not (i would guess it actually doesn't normally), but looking again, this is ok

jreback commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -395,6 +403,163 @@ def f(x):
+    elif is_integer_dtype(data):
+        # treat as multiples of the given unit
+        data, copy_made = ints_to_td64ns(data, unit=unit)
+        copy = copy and not copy_made
```
do u actually test this is not making a copy? on int64 input is the only case i think

jbrockmendel replied:
I know I've seen tests about similar things, but not entirely sure they hit this case.  I think I'll implement a test and in one of the upcoming test-collection passes keep an eye out for similar tests to group together.

jreback replied:
ok sure, yeah keeping zero copy is hard, we have very limited testing on this now

jreback commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -395,6 +403,163 @@ def f(x):
+        dtype_str = "timedelta64[{unit}]".format(unit=unit)
+        data = data.view(dtype_str)
+
+        # TODO: watch out for overflows when converting from lower-resolution
```
can this be just a view?

jbrockmendel replied:
No, that incorrectly multiplies/divides the values:

```
>>> arr = np.array([1], 'm8[s]')
>>> arr.view('m8[ns]')
array([1], dtype='timedelta64[ns]')
```

jreback commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -395,6 +403,163 @@ def f(x):
 # ---------------------------------------------------------------------
 # Constructor Helpers
 
+def sequence_to_td64ns(data, copy=False, unit='ns', errors='raise'):
```
you use single and double quotes for args inconsistently -

jreback commented on pandas/core/indexes/timedeltas.py:
```diff
@@ -154,32 +153,73 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
-            raise ValueError('TimedeltaIndex() must be called with a '
+            raise ValueError('{cls}() must be called with a '
                              'collection of some kind, {data} was passed'
-                             .format(data=repr(data)))
```
same

jreback commented on pandas/tests/indexes/timedeltas/test_arithmetic.py:
```diff
@@ -453,10 +453,16 @@ def test_timedelta_ops_with_missing_values(self):
-        sn = pd.to_timedelta(Series([pd.NaT]))
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
+            # Passing datetime64-dtype data to TimedeltaIndex is deprecated
+            sn = pd.to_timedelta(Series([pd.NaT]))
```
this is not correct 
it is all nan which is not definitive

jbrockmendel replied:
`Series([pd.NaT])` has `datetime64[ns]` dtype.  After we've done inference, we need to stick with it.  I can make this test clearer by putting the dtype in the Series constructor explicitly

jreback replied:
what i mean is that this is all nan and inference is ambiguous at this point. so I wouldn't expect the warning (unless the dtype IS explicity passed). Can address in a followup, but pls open an issue, this is IMHO a bug.

jreback commented on pandas/tests/indexes/timedeltas/test_arithmetic.py:
```diff
@@ -453,10 +453,16 @@ def test_timedelta_ops_with_missing_values(self):
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
+            # Passing datetime64-dtype data to TimedeltaIndex is deprecated
+            dfn = pd.DataFrame([pd.NaT]).apply(pd.to_timedelta)
+
```
same

do u actually have a case that tests the warnings?

jbrockmendel replied:
Will implement.

jbrockmendel replied:
actually we have it in test_constructors.test_dt64_data_invalid

jreback commented on pandas/tests/indexes/timedeltas/test_construction.py:
```diff
@@ -5,11 +5,74 @@
+        # GH#23539
+        # fast-path for inferring a frequency if the passed data already
+        #  has one
+        tdi = pd.timedelta_range('1 second', periods=10**7, freq='1s')
```
we do not have any existing tests for this?

jbrockmendel replied:
nope

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into pre-fixes

jbrockmendel committed:
change ValueError-->TypeError

jbrockmendel committed:
double quotes

jbrockmendel committed:
test that no copy is made with int64 data

jbrockmendel committed:
update tests to TypeError

jreback commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+
+    Returns
+    -------
+    ndarray[timedelta64[ns]]
```
is this how we indicate a multiple returns here? @datapythonista

datapythonista replied:
The right format IMO should be something like:
```
Returns
-------
data : numpy.ndarray
    The sequence converted to a numpy array of type ``timedelta64[ns]``.
inferred_freq : Tick or None
    The inferred frequency of the sequence.
```

We're adding this validation in #23432

jreback commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+        data = data._data
+
+    # Convert whatever we have into timedelta64[ns] dtype
+    if is_object_dtype(data) or is_string_dtype(data):
```
why is the is_string_dtype needed?

jbrockmendel replied:
because we have test cases where we pass e.g. `np.array(['2 Days', '3 Days'], dtype='|S6')`

jreback commented on pandas/core/tools/timedeltas.py:
```diff
@@ -129,31 +126,27 @@ def _convert_listlike(arg, unit='ns', box=True, errors='raise', name=None):
+        #  returning arg (errors == "ignore"), and where the input is a
+        #  generator, we return a useful list-like instead of a
+        #  used-up generator
         arg = np.array(list(arg), dtype='O')
```
-> object. We are not consistent across the codebase with specifiying object dtype (not a big deal as its a small thing), but small things add up

jbrockmendel replied:
done

jreback reviewed (changes requested):
just some minor comments.

jreback commented:
@gfyoung if you have any further comments.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into pre-fixes

jbrockmendel committed:
dtype=object instead of 'O'

jbrockmendel commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -969,6 +970,7 @@ Deprecations
 - The class ``FrozenNDArray`` has been deprecated. When unpickling, ``FrozenNDArray`` will be unpickled to ``np.ndarray`` once this class is removed (:issue:`9031`)
 - Deprecated the `nthreads` keyword of :func:`pandas.read_feather` in favor of
   `use_threads` to reflect the changes in pyarrow 0.11.0. (:issue:`23053`)
+- Constructing a :class:`TimedeltaIndex` from data with ``datetime64``-dtyped data is deprecated, will raise ``TypeError`` in a future version (:issue:`23539`)
```
as of just now, yes

jreback commented:
@jbrockmendel ok let's merge this on green. a couple of minor followups (they are unresolved conversations).

jbrockmendel committed:
use pytest.raises instead of tm.assert_raises_regex

datapythonista commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+    array : list-like
+    copy : bool, default False
+    unit : str, default "ns"
+    errors : {"raise", "coerce", "ignore"}, default "raise"
```
If this is public, it'd be nice to have a short description for the parameters.

datapythonista commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+
+    Raises
+    ------
+    ValueError : data cannot be converted to timedelta64[ns]
```
If you can capitalize the first letter of the sentence, and finish with a period, that would be great.
```suggestion
    ValueError : Data cannot be converted to timedelta64[ns].
```

datapythonista commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+
+    Notes
+    -----
+    Unlike `pandas.to_timedelta`, if setting `errors=ignore` will not cause
```
Still under discussion, but I think the standard we're moving to is to use single backticks for variables, arguments, attributes, functions, methods classes. And double backticks for values and code or pseudo-code. I think the `errors=ignore` should be double backticks.
```suggestion
    Unlike `pandas.to_timedelta`, if setting ```errors=ignore``` will not cause
```

datapythonista commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+    Parameters
+    ----------
+    data : np.ndarray with integer-dtype
+    unit : str, default "ns"
```
Short description of params.

Also, I'd move the `with integer-dtype` to the description, and I wouldn't abbreviate numpy in the type (I wouldn't anywhere, and I don't know if there is agreement on not abbreviating it here).

At some point would be nice to have a limited list of types (`str, int, list-like, numpy.array, Series...`), and that would allow us to validate typos, descriptions instead of types... So, would be nice to have just pure types here, and the additional information in the description.

jbrockmendel replied:
Good notes all around.  Mind if I apply them in the next pass?  The CI is pretty backed up ATM.

datapythonista replied:
Sure. We'll start soon adding them to the CI and that won't be an option, but you're still on time to do it in two passes. :)

datapythonista commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+    Returns
+    -------
+    ndarray[timedelta64[ns]]
+    bool : whether a copy was made
```
Same format as before (same as in parameters)

datapythonista commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+    ----------
+    data : ndarray or Index
+    unit : str, default "ns"
+    errors : {"raise", "coerce", "ignore"}, default "raise"
```
same

datapythonista commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+
+    Returns
+    -------
+    ndarray[timedelta64[ns]]
```
`numpy.ndarray` and a short description including the type.

datapythonista commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -397,6 +405,163 @@ def f(x):
+
+    Raises
+    ------
+    ValueError : data cannot be converted to timedelta64[ns]
```
Capital + period

datapythonista reviewed (commented):
Added some suggestions and info about the docstrings format.

jbrockmendel commented:
@datapythonista I'm adding suggested edits into #23587 (needs rebasing anyway).  Definitely good suggestions.

For the "unit" description the best I've come up with is "The timedelta unit to treat integers as multiples of"; I can't think of a natural-sounding way to rephrase this that doesn't end in a preposition.  Thoughts?

datapythonista commented:
That description sounds good to me. Can't think of anything better right now.

jorisvandenbossche commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -1054,11 +1054,11 @@ def test_tdi_mul_float_series(self, box_df_fail):
         box2 = pd.Series if box is pd.Index else box
         expected = tm.box_expected(expected, box2)
 
-        result = idx * Series(rng5f + 0.1)
```
@jbrockmendel why did you change this? Left-over from initially changing the behaviour on floats?

jbrockmendel replied:
Yes.
